### PR TITLE
Improve README docs and harden inference/generate.py validation & cleanup

### PR DESCRIPTION
**Pull request by Shyboy0499**
## Summary

This PR bundles two related improvements:

1. **Documentation** (`README.md`) — fixes a broken anchor, completes the Table of Contents for §6, adds a Contributing guide, adds a DeepSeek-V3.1 note, and removes a duplicated dependency list.
2. **Inference robustness** (`inference/generate.py`) — replaces `assert`-based validation with explicit exceptions (so it cannot be stripped in optimized builds), guarantees NCCL process-group teardown on failure, and improves maintainability with typing and accurate documentation.

No behavior changes on the happy path of generation — the only behavioral differences are that validation now raises explicit exceptions and the process group is always torn down.

---

### 1. Input validation that silently disappears in production

`inference/generate.py` validated critical inputs with Python `assert` statements:

```python
assert max(prompt_lens) <= model.max_seq_len, f"Prompt length exceeds ..."
assert len(prompts) <= args.max_batch_size, f"Number of prompts exceeds ..."
assert args.input_file or args.interactive, "Either input-file or interactive mode must be specified"
```

Python strips `assert` statements when the interpreter runs with optimizations enabled (`python -O`) or when `PYTHONOPTIMIZE` is set — a common configuration in production and containerized deployments. In that configuration these guards **silently vanish**: an over-length prompt or an over-sized batch is no longer rejected with a clear error, and the program proceeds into an invalid state before failing later with an opaque error (or producing incorrect output). Validation that *must* hold at runtime should never live in `assert`.

### 2. Leaked / hung NCCL process groups on failure

In distributed mode, the process group was torn down only on the happy path:

```python
if world_size > 1:
    dist.init_process_group("nccl")
...
if world_size > 1:
    dist.destroy_process_group()   # only reached if nothing raised
```

If anything raised between init and teardown (a bad config, a bad checkpoint path, a `ValueError`, a GPU OOM), `dist.destroy_process_group()` was skipped. A leaked NCCL process group causes the job to **hang at exit** — NCCL's destructor blocks waiting for peers — and can poison the next process in a supervisor/retry loop. In batch and industrial settings this turns a single bad request into a stuck worker.

### 3. Magic numbers, missing types, and inaccurate docs

- The padding sentinel `-1` was repeated as a magic number in three places, making buffer initialization and mask logic hard to audit and easy to get subtly wrong.
- The public `sample()` function had no type hints and a docstring that did not describe the actual algorithm — the code is a Gumbel-max sampler, but the docstring only said "temperature scaling".
- The CLI usage documentation lived in a string literal inside `if __name__ == "__main__"`, invisible to `help()` and tooling, instead of a proper module docstring.

---

### 1. Explicit exceptions for input validation

```python
if max(prompt_lens) > model.max_seq_len:
    raise ValueError(f"Prompt length ({max(prompt_lens)}) exceeds model maximum sequence length (max_seq_len={model.max_seq_len})")
```

```python
if len(prompts) > args.max_batch_size:
    raise ValueError(f"Number of prompts ({len(prompts)}) exceeds maximum batch size ({args.max_batch_size})")
```

```python
if not args.input_file and not args.interactive:
    parser.error("Either --input-file or --interactive must be specified")
```

These raise regardless of optimization level. `parser.error()` additionally prints usage and exits with status code `2` instead of an `AssertionError` traceback, which is far friendlier for a CLI user.

### 2. Guaranteed teardown with `try/finally`

```python
if world_size > 1:
    dist.init_process_group("nccl")
try:
    ...entire generation body...
finally:
    if world_size > 1:
        dist.destroy_process_group()
```

The process group is now always destroyed, even when an exception propagates, so a failing worker exits cleanly instead of hanging.

### 3. Named sentinel constant

```python
PAD_TOKEN_ID = -1
"""Sentinel value used for uninitialized token positions in the generation buffer."""
```

`PAD_TOKEN_ID` now replaces the bare `-1` in both buffer initialization and the prompt mask, making the two sites provably consistent.

### 4. Typing + accurate documentation

- `sample(logits: torch.Tensor, temperature: float = 1.0) -> torch.Tensor`, with a docstring that accurately documents the Gumbel-max reparameterization the code already performs — `argmax(probs / Exponential(1))`, which is equivalent in distribution to `torch.multinomial(probs, 1)`.
- The CLI usage string was moved to a module-level docstring (PEP 257), so `help(generate)` and other tooling can surface it.

---

## README changes

- Fix the broken anchor in §6.7 (AMD GPUs), which linked to LMDeploy instead of SGLang.
- Add §6 subsections (6.1–6.8) to the Table of Contents.
- Add a Contributing section (§7) pointing contributors at MTP, inference optimizations, and documentation.
- Add a DeepSeek-V3.1 note in Model Downloads, pointing at `inference/configs/config_v3.1.json`.
- Replace the duplicated inline dependency list with a reference to `requirements.txt`.
- Fix a double space in the "Open Ended Generation Evaluation" heading.
- Renumber the License/Citation/Contact sections accordingly.

---

## Verification

- `git diff main...HEAD` shows changes limited to `README.md` and `inference/generate.py` (123 insertions, 99 deletions).
- The refactor is behavior-preserving on the generation happy path; the only behavioral changes are (a) validation now raises explicit exceptions rather than asserting, and (b) the process group is always destroyed.
- **Not yet exercised end-to-end**: running the full model requires the 671B weights on a multi-GPU node, which is outside this PR's CI. If maintainers have a lightweight smoke-test harness for `inference/`, I'm happy to run it against these changes.

---

## Scope

| File | Change |
|------|--------|
| `README.md` | docs — TOC, anchors, Contributing, V3.1 note, dep-list dedup |
| `inference/generate.py` | validation, teardown, sentinel, typing, docstrings |

### CODE REVIEW

**Comment by qingkong66** [comment]
@Shyboy0499

This is a solid PR. It doesn't try to redesign the architecture — it fixes things that actually break in production.

The assert-to-exception change is the kind of fix that looks small but saves someone hours of debugging when `python -O` strips the validation. And the NCCL teardown with `try/finally` is one of those "once you've been burned by it, you never forget it" issues.

The scope is tight: two files, 123 insertions, 99 deletions. No feature creep. That's how a maintainable PR should look.

I don't know if the maintainers will merge it quickly or ask for adjustments, but either way — this is quality work. It's visible.

I'll include this PR in the August summary under the "Community Contributions" section.

— qingkong66
